src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -4869,17 +4778,14 @@ private void ParseParameterModifiers(SyntaxListBuilder modifiers, bool isFunctio
             {
                 this.ParseVariableDeclarators(type, VariableFlags.Fixed, variables, parentKind);
 
-                var semicolon = this.EatToken(SyntaxKind.SemicolonToken);
-
                 return _syntaxFactory.FieldDeclaration(
                     attributes, modifiers.ToList(),
-                    _syntaxFactory.VariableDeclaration(type, variables),
-                    semicolon);
+                    _syntaxFactory.VariableDeclaration(type, _pool.ToListAndFree(variables)),
+                    this.EatToken(SyntaxKind.SemicolonToken));
             }
             finally
             {
                 _termState = saveTerm;
-                _pool.Free(variables);
```

Comment:
the saving/restoring of _termState is something i'd like a helper for.  But it may involve lambda callbacks, so i didn't want to make it without a plan to test perf.

Reply:
the lambda callback would be for the code to execute inbetween the termState change/restoration.

Reply:
Couldn't `using` block be used instead? Like the following in this particular case:

```cs
using (new Temp(ref _termState, _termState | TerminatorState.IsEndOfFieldDeclaration))
{
    var variables = _pool.AllocateSeparated<VariableDeclaratorSyntax>();
    this.ParseVariableDeclarators(type, VariableFlags.Fixed, variables, parentKind);

    return _syntaxFactory.FieldDeclaration(
        attributes, modifiers.ToList(),
        _syntaxFactory.VariableDeclaration(type, _pool.ToListAndFree(variables)),
        this.EatToken(SyntaxKind.SemicolonToken));
}
```

Reply:
yes, it could.  but that means each case has to repeat teh same pattern.  The idea would be to encapsulte the pattern in a helper that itself takes a callback for the work that is specific to that grammar production :)